src/state_causal_type.erl

```diff
@@ -0,0 +1,275 @@
+                                                             Object,
+                                                             DataStore)}
+                             end
+                     end, InterDataStore, DataStoreA),
+    MergedDataStore1 =
+        orddict:fold(fun(DotContext, Object, {dot_fun, DataStore}) ->
+                             case ordsets:is_element(DotContext, DotCloudA) of
+                                 true ->
+                                     {dot_fun, DataStore};
+                                 false ->
+                                     {dot_fun, orddict:store(DotContext,
+                                                             Object,
+                                                             DataStore)}
+                             end
+                     end, MergedDataStore0, DataStoreB),
+    {MergedDataStore1, ordsets:union(DotCloudA, DotCloudB)};
+causal_join({{{dot_map, ValueDataStoreType}, DataStoreA}=Fst, DotCloudA},
+            {{{dot_map, ValueDataStoreType}, DataStoreB}=Snd, DotCloudB}) ->
+    UnionObjects = ordsets:union(ordsets:from_list(orddict:fetch_keys(DataStoreA)),
+                                 ordsets:from_list(orddict:fetch_keys(DataStoreB))),
+    MergedDataStore =
+        ordsets:fold(fun(Object, MergedDataStore0) ->
+                             {ok, SubDataStoreA} = get_sub_data_store(Object, Fst),
+                             {ok, SubDataStoreB} = get_sub_data_store(Object, Snd),
+                             {MergedSubDataStore, _} =
+                                 causal_join({SubDataStoreA, DotCloudA},
+                                             {SubDataStoreB, DotCloudB}),
+                             case is_bottom_data_store(MergedSubDataStore) of
+                                 true ->
+                                     MergedDataStore0;
+                                 false ->
+                                     insert_object(Object,
+                                                   MergedSubDataStore,
+                                                   MergedDataStore0)
+                             end
+                     end, new_data_store({dot_map, ValueDataStoreType}), UnionObjects),
+    {MergedDataStore, ordsets:union(DotCloudA, DotCloudB)}.
+
+%% @doc Determine if a change for a given causal crdt is an inflation or not.
+%%
+%% Given a particular causal crdt and two instances of that causal crdt,
+%% determine if `B(second)' is an inflation of `A(first)'.
+-spec is_lattice_inflation(causal_crdt(), causal_crdt()) -> boolean().
+is_lattice_inflation({DataStoreA, DotCloudA}, {DataStoreB, DotCloudB}) ->
+    DotCloudAList = get_maxs_for_all(DotCloudA),
+    DotCloudBList = get_maxs_for_all(DotCloudB),
+    is_inflation(DotCloudAList, DotCloudBList) andalso
+        ordsets:is_subset(get_dot_cloud(DataStoreB), get_dot_cloud(DataStoreA)).
+
+%% @doc Create an empty DataStore.
```

Comment:
It looks like this function `new_data_store` could be used in the `create_causal_crdt` function to reduce duplication.